File: cpp/821.shortest-distance-to-a-character.cpp
#include <algorithm>
#include <climits>
#include <string>
#include <vector>

using std::string;
using std::vector;

class Solution {
public:
  vector<int> shortestToChar(string s, char c) {
    vector<int> res(s.size(), s.size());

    int pos = s.size();

    for (int i = 0; i < s.size(); i++) {
      if (s[i] == c) {
        pos = i;
      }
      res[i] = abs(i - pos);
    }

    for (int i = s.size() - 1; i >= 0; i--) {
      if (s[i] == c) {
        pos = i;
      }
      res[i] = std::min(res[i], abs(pos - i));
    }

    return res;
  }
};

File: python/821.shortest-distance-to-a-character.py
class Solution:
    def shortestToChar(self, s: str, c: str) -> list[int]:
        n = len(s)

        res = [n] * n
        pos = -n

        for i in range(n):
            if s[i] == c:
                pos = i

            res[i] = i - pos

        for i in range(n - 1, -1, -1):
            if s[i] == c:
                pos = i

            res[i] = min(res[i], abs(pos - i))

        return res

